### PR TITLE
Migrated `pkg/scheduler/framework/preemption & defaultpreemption` to use contextual logging

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Migrated the preemption scheduler plugin to use contextual logging

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1467645065

#### Special notes for your reviewer:
 the test file `pkg/scheduler/framework/preemption/preemption_test.go` is updated in PR #116842 https://github.com/kubernetes/kubernetes/pull/116842/files#diff-33df49b5fb3da50ca769dc3b9fc05cc3f55a0f7cd8877eba9047fa1a504f52de
#### Does this PR introduce a user-facing change?

```release-note
Migrated `pkg/scheduler/framework/preemption` to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

/sig instrumentation
/wg structured-logging
/area logging
/sig scheduling

### CODE REVIEW

File: pkg/scheduler/framework/preemption/preemption.go [hunk 1]
```diff
@@ -234,12 +237,12 @@ func (ev *Evaluator) findCandidates(ctx context.Context, pod *v1.Pod, m framewor
 	}
 
 	offset, numCandidates := ev.GetOffsetAndNumCandidates(int32(len(potentialNodes)))
-	if klogV := klog.V(5); klogV.Enabled() {
+	if logger := logger.V(5); logger.Enabled() {
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/framework/preemption/preemption.go:240]: Nope, this isn't klog anymore. `logger` and `loggerV` inside such an `if` statement are the recommended names for a `go-logr/logr.Logger` instance.

klog is merely a wrapper around that.

[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/preemption/preemption.go:240]: Then loggerV

File: pkg/scheduler/framework/preemption/preemption.go [hunk 2]
```diff
@@ -343,6 +346,7 @@ func (ev *Evaluator) SelectCandidate(candidates []Candidate) Candidate {
 // - Reject the victim pods if they are in waitingPod map
 // - Clear the low-priority pods' nominatedNodeName status if needed
 func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.Pod, pluginName string) *framework.Status {
+	logger := klog.FromContext(ctx)
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go]: Done.

**Comment by mengjiao-liu** [comment]
Migrate PR https://github.com/kubernetes/kubernetes/pull/116633 to this PR based on this comment https://github.com/kubernetes/kubernetes/pull/116633/files#r1152703840.

File: pkg/scheduler/framework/preemption/preemption.go [hunk 3]
```diff
@@ -272,7 +275,7 @@ func (ev *Evaluator) callExtenders(pod *v1.Pod, candidates []Candidate) ([]Candi
 		nodeNameToVictims, err := extender.ProcessPreemption(pod, victimsMap, nodeLister)
 		if err != nil {
 			if extender.IsIgnorable() {
-				klog.InfoS("Skipping extender as it returned error and has ignorable flag set",
+				logger.Info("Skipping extender as it returned error and has ignorable flag set",
```
Comment (alculquicondor) [inline pkg/scheduler/framework/preemption/preemption.go:278]:
```suggestion
				logger.Info("Skipped extender as it returned error and has ignorable flag set",
```

File: pkg/scheduler/framework/preemption/preemption.go [hunk 4]
```diff
@@ -367,13 +371,13 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
...
 				if _, err := cs.CoreV1().Pods(victim.Namespace).ApplyStatus(ctx, victimPodApply, metav1.ApplyOptions{FieldManager: fieldManager, Force: true}); err != nil {
-					klog.ErrorS(err, "Preparing pod preemption", "pod", klog.KObj(victim), "preemptor", klog.KObj(pod))
+					logger.Error(err, "Preparing pod preemption", "pod", klog.KObj(victim), "preemptor", klog.KObj(pod))
 					errCh.SendErrorWithCancel(err, cancel)
 					return
 				}
 			}
 			if err := util.DeletePod(ctx, cs, victim); err != nil {
-				klog.ErrorS(err, "Preempting pod", "pod", klog.KObj(victim), "preemptor", klog.KObj(pod))
+				logger.Error(err, "Preempting pod", "pod", klog.KObj(victim), "preemptor", klog.KObj(pod))
```
Comment (alculquicondor) [inline pkg/scheduler/framework/preemption/preemption.go:380]:
```suggestion
				logger.Error(err, "Preempted pod", "pod", klog.KObj(victim), "preemptor", klog.KObj(pod))
```

**Review by alculquicondor — COMMENTED** [review]
Some performance concern and general housekeeping based on https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#message-style-guidelines

File: pkg/scheduler/framework/preemption/preemption.go [hunk 5]
```diff
@@ -272,7 +275,7 @@ func (ev *Evaluator) callExtenders(pod *v1.Pod, candidates []Candidate) ([]Candi
 		nodeNameToVictims, err := extender.ProcessPreemption(pod, victimsMap, nodeLister)
 		if err != nil {
 			if extender.IsIgnorable() {
-				klog.InfoS("Skipping extender as it returned error and has ignorable flag set",
+				logger.Info("Skipping extender as it returned error and has ignorable flag set",
 					"extender", extender, "err", err)
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go:279]: It seems the name is enough.

File: pkg/scheduler/framework/preemption/preemption.go [hunk 6]
```diff
@@ -234,12 +237,12 @@ func (ev *Evaluator) findCandidates(ctx context.Context, pod *v1.Pod, m framewor
...
 
 	offset, numCandidates := ev.GetOffsetAndNumCandidates(int32(len(potentialNodes)))
-	if klogV := klog.V(5); klogV.Enabled() {
+	if loggerV := logger.V(5); logger.Enabled() {
 		var sample []string
 		for i := offset; i < offset+10 && i < int32(len(potentialNodes)); i++ {
 			sample = append(sample, potentialNodes[i].Node().Name)
 		}
-		klogV.InfoS("Selecting candidates from a pool of nodes", "potentialNodesCount", len(potentialNodes), "offset", offset, "sampleLength", len(sample), "sample", sample, "candidates", numCandidates)
+		loggerV.Info("Selecting candidates from a pool of nodes", "potentialNodesCount", len(potentialNodes), "offset", offset, "sampleLength", len(sample), "sample", sample, "candidates", numCandidates)
```
Comment (kerthcet) [inline pkg/scheduler/framework/preemption/preemption.go:245]: Also as suggested:
```suggestion
		loggerV.Info("Selected candidates from a pool of nodes", "potentialNodesCount", len(potentialNodes), "offset", offset, "sampleLength", len(sample), "sample", sample, "candidates", numCandidates)
```

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 7]
```diff
@@ -207,7 +207,7 @@ func (pl *DefaultPreemption) SelectVictimsOnNode(
 			}
 			rpi := pi.Pod
 			victims = append(victims, rpi)
-			klog.V(5).InfoS("Pod is a potential preemption victim on node", "pod", klog.KObj(rpi), "node", klog.KObj(nodeInfo.Node()))
+			klog.FromContext(ctx).V(5).Info("Pod is a potential preemption victim on node", "pod", klog.KObj(rpi), "node", klog.KObj(nodeInfo.Node()))
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go]: Done. Nice catch！

File: pkg/scheduler/framework/preemption/preemption.go [hunk 8]
```diff
@@ -150,20 +150,22 @@ type Evaluator struct {
...
+
 	// 0) Fetch the latest version of <pod>.
 	// It's safe to directly fetch pod here. Because the informer cache has already been
 	// initialized when creating the Scheduler obj.
 	// However, tests may need to manually initialize the shared pod informer.
 	podNamespace, podName := pod.Namespace, pod.Name
 	pod, err := ev.PodLister.Pods(pod.Namespace).Get(pod.Name)
 	if err != nil {
-		klog.ErrorS(err, "Getting the updated preemptor pod object", "pod", klog.KRef(podNamespace, podName))
+		logger.Error(err, "Getting the updated preemptor pod object", "pod", klog.KRef(podNamespace, podName))
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go]: Done.

File: pkg/scheduler/framework/preemption/preemption.go [hunk 9]
```diff
@@ -217,12 +219,13 @@ func (ev *Evaluator) findCandidates(ctx context.Context, pod *v1.Pod, m framewor
...
 	}
+	logger := klog.FromContext(ctx)
 	potentialNodes, unschedulableNodeStatus := nodesWherePreemptionMightHelp(allNodes, m)
 	if len(potentialNodes) == 0 {
-		klog.V(3).InfoS("Preemption will not help schedule pod on any node", "pod", klog.KObj(pod))
+		logger.V(3).Info("Preemption will not help schedule pod on any node", "pod", klog.KObj(pod))
 		// In this case, we should clean-up any existing nominated node name of the pod.
 		if err := util.ClearNominatedNodeName(ctx, ev.Handler.ClientSet(), pod); err != nil {
-			klog.ErrorS(err, "Cannot clear 'NominatedNodeName' field of pod", "pod", klog.KObj(pod))
+			logger.Error(err, "Cannot clear 'NominatedNodeName' field of pod", "pod", klog.KObj(pod))
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go]: Done.

File: pkg/scheduler/framework/preemption/preemption.go (hunk 3 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go]: Done.

File: pkg/scheduler/framework/preemption/preemption.go [hunk 10]
```diff
@@ -367,13 +371,13 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
 				)
 
 				if _, err := cs.CoreV1().Pods(victim.Namespace).ApplyStatus(ctx, victimPodApply, metav1.ApplyOptions{FieldManager: fieldManager, Force: true}); err != nil {
-					klog.ErrorS(err, "Preparing pod preemption", "pod", klog.KObj(victim), "preemptor", klog.KObj(pod))
+					logger.Error(err, "Preparing pod preemption", "pod", klog.KObj(victim), "preemptor", klog.KObj(pod))
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go]: Done.

File: pkg/scheduler/framework/preemption/preemption.go (hunk 4 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go]: Done.

File: pkg/scheduler/framework/preemption/preemption.go (hunk 6 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go]: Done.

File: pkg/scheduler/framework/preemption/preemption.go [hunk 11]
```diff
@@ -283,7 +286,7 @@ func (ev *Evaluator) callExtenders(pod *v1.Pod, candidates []Candidate) ([]Candi
 			if victims == nil || len(victims.Pods) == 0 {
 				if extender.IsIgnorable() {
 					delete(nodeNameToVictims, nodeName)
-					klog.InfoS("Ignoring node without victims", "node", klog.KRef("", nodeName))
+					logger.Info("Ignoring node without victims", "node", klog.KRef("", nodeName))
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption.go]: Just log the name.

**Comment by mengjiao-liu** [comment]
All comments addressed.

**Review by alculquicondor — COMMENTED** [review]
/approve

**Pull request merged**